File: tbt-bridge/src/com/pulsar/ns400z/tbtbridge/PulsarForegroundService.java
package com.pulsar.ns400z.tbtbridge;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ServiceInfo;
import android.os.Build;
import android.os.IBinder;
import android.util.Log;

/**
 * Android 14/15 Compliant Persistent Foreground Service (FOREGROUND_SERVICE_TYPE_CONNECTED_DEVICE).
 * Keeps Bluetooth GATT connection alive, coordinates telemetry, media, and navigation pipelines.
 */
public class PulsarForegroundService extends Service implements PulsarBleManager.BleListener {

    private static final String TAG = "PulsarForegroundService";
    public static final String CHANNEL_ID = "pulsar_connect_live_channel";
    public static final int NOTIFICATION_ID = 4001;

    public static final String ACTION_START = "com.pulsar.ns400z.tbtbridge.START_SERVICE";
    public static final String ACTION_STOP = "com.pulsar.ns400z.tbtbridge.STOP_SERVICE";
    public static final String ACTION_CONNECT = "com.pulsar.ns400z.tbtbridge.CONNECT_BLE";

    private static PulsarForegroundService instance;

    private PulsarBleManager bleManager;
    private MediaStateListener mediaListener;
    private TelephonyCallHandler callHandler;
    private PhoneStateMonitor phoneMonitor;

    private boolean isConnected = false;
    private String activeManeuver = "Navigation Idle";
    private String activeStreet = "--";

    public static PulsarForegroundService getInstance() {
        return instance;
    }

    public static void start(Context context) {
        Intent intent = new Intent(context, PulsarForegroundService.class);
        intent.setAction(ACTION_START);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.startForegroundService(intent);
        } else {
            context.startService(intent);
        }
    }

    @Override
    public void onCreate() {
        super.onCreate();
        instance = this;
        Log.i(TAG, "PulsarForegroundService onCreate()");

        createNotificationChannel();
        startForegroundWithNotification("Initializing Pulsar Connect...", "Connecting to NS400Z Cluster...");

        bleManager = PulsarBleManager.getInstance(this);
        bleManager.addListener(this);

        mediaListener = new MediaStateListener(this, bleManager);
        callHandler = new TelephonyCallHandler(this, bleManager);
        phoneMonitor = new PhoneStateMonitor(this, bleManager, callHandler);
        phoneMonitor.start();

        // Connect to motorcycle cluster
        bleManager.startScanOrConnect();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null && intent.getAction() != null) {
            String action = intent.getAction();
            if (ACTION_STOP.equals(action)) {
                stopForeground(true);
                stopSelf();
                return START_NOT_STICKY;
            } else if (ACTION_CONNECT.equals(action)) {
                bleManager.startScanOrConnect();
            }
        }
        return START_STICKY;
    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                    CHANNEL_ID,
                    "Pulsar NS400Z Live Service",
                    NotificationManager.IMPORTANCE_LOW
            );
            channel.setDescription("Background Bluetooth Link and Real-time Navigation Stream");
            channel.setShowBadge(false);
            NotificationManager nm = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
            if (nm != null) {
                nm.createNotificationChannel(channel);
            }
        }
    }

    private void startForegroundWithNotification(String title, String content) {
        Notification notification = buildNotification(title, content);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            startForeground(NOTIFICATION_ID, notification, ServiceInfo.FOREGROUND_SERVICE_TYPE_CONNECTED_DEVICE);
        } else {
            startForeground(NOTIFICATION_ID, notification);
        }
    }

    private Notification buildNotification(String title, String content) {
        Intent launchIntent = new Intent(this, MainActivity.class);
        PendingIntent pi = PendingIntent.getActivity(
                this, 0, launchIntent,
                PendingIntent.FLAG_UPDATE_CURRENT | (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M ? PendingIntent.FLAG_IMMUTABLE : 0)
        );

        Notification.Builder builder;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            builder = new Notification.Builder(this, CHANNEL_ID);
        } else {
            builder = new Notification.Builder(this);
        }

        builder.setContentTitle(title)
                .setContentText(content)
                .setSmallIcon(android.R.drawable.ic_dialog_map)
                .setContentIntent(pi)
                .setOngoing(true);

        return builder.build();
    }

    private void updateNotification() {
        String title = isConnected ? "Pulsar NS400Z: Connected" : "Pulsar NS400Z: Disconnected";
        String content = "Nav: " + activeManeuver + " (" + activeStreet + ")";
        if (mediaListener != null && mediaListener.isPlaying()) {
            content += " | ♫ " + mediaListener.getCurrentTitle();
        }

        NotificationManager nm = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        if (nm != null) {
            nm.notify(NOTIFICATION_ID, buildNotification(title, content));
        }
    }

    public void updateNavStatus(String maneuver, String street) {
        this.activeManeuver = maneuver;
        this.activeStreet = street;
        updateNotification();
    }

    @Override
    public void onConnectionStateChanged(boolean connected, String deviceName, String deviceAddress) {
        this.isConnected = connected;
        updateNotification();

        // Frames sent while disconnected are dropped by PulsarBleManager, and the
        // cluster loses its now-playing state on every power-cycle/reconnect.
        // Re-push the current media state as soon as the link is up. This fires
        // twice per connection (pre- and post-service-discovery); the first call
        // is dropped harmlessly, the second lands once 0610 is bound.
        if (connected && mediaListener != null) {
            mediaListener.syncMetadata();
        }
    }

    @Override
    public void onPacketSent(String charUuid, byte[] frame, boolean success) {
        // Telemetry/TBT dispatched
    }

    @Override
    public void onHandlebarEvent(PulsarProtocol.HandlebarEvent event) {
        Log.i(TAG, "Handlebar switch event received from cluster!");
        if (mediaListener != null) {
            mediaListener.handleHandlebarMedia(event);
        }
        if (callHandler != null) {
            callHandler.handleHandlebarCallAction(event);
        }
    }

    public void refreshMediaSessions() {
        if (mediaListener != null) {
            mediaListener.initMediaSessions();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        instance = null;
        Log.i(TAG, "PulsarForegroundService onDestroy()");
        if (phoneMonitor != null) {
            phoneMonitor.stop();
        }
        if (bleManager != null) {
            bleManager.removeListener(this);
            bleManager.disconnect();
        }
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: app/src/com/bajaj/rideconnect/re/PulsarForegroundService.kt
package com.bajaj.rideconnect.re

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.pm.ServiceInfo
import android.os.Build
import android.os.IBinder
import android.util.Log

class PulsarForegroundService : Service(), PulsarBleManager.BleListener {

    private var bleManager: PulsarBleManager? = null
    var mediaListener: MediaStateListener? = null
        private set
    private var callHandler: TelephonyCallHandler? = null
    private var phoneMonitor: PhoneStateMonitor? = null

    private var isConnected = false
    private var activeManeuver = "Navigation Idle"
    private var activeStreet = "--"

    override fun onCreate() {
        super.onCreate()
        instance = this
        Log.i(TAG, "PulsarForegroundService onCreate()")

        createNotificationChannel()
        startForegroundWithNotification("Initializing Pulsar Connect...", "Connecting to NS400Z Cluster...")

        val ble = PulsarBleManager.getInstance(this)
        this.bleManager = ble
        ble.addListener(this)

        val media = MediaStateListener(this, ble)
        this.mediaListener = media
        val call = TelephonyCallHandler(this, ble)
        this.callHandler = call
        val monitor = PhoneStateMonitor(this, ble, call)
        this.phoneMonitor = monitor
        monitor.start()

        ble.startScanOrConnect()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        val action = intent?.action
        if (action != null) {
            when (action) {
                ACTION_STOP -> {
                    bleManager?.apply {
                        setAutoReconnect(false)
                        stopScan()
                        disconnect()
                    }
                    phoneMonitor?.stop()
                    val nm = getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager
                    nm?.apply {
                        cancel(NOTIFICATION_ID)
                        cancelAll()
                    }
                    stopForeground(STOP_FOREGROUND_REMOVE)
                    stopSelf()
                    return START_NOT_STICKY
                }
                ACTION_CONNECT -> {
                    bleManager?.startScanOrConnect()
                }
            }
        }
        return START_STICKY
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                "Pulsar NS400Z Live Service",
                NotificationManager.IMPORTANCE_LOW
            ).apply {
                description = "Background Bluetooth Link and Real-time Navigation Stream"
                setShowBadge(false)
            }
            val nm = getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager
            nm?.createNotificationChannel(channel)
        }
    }

    private fun startForegroundWithNotification(title: String, content: String) {
        val notification = buildNotification(title, content)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            startForeground(NOTIFICATION_ID, notification, ServiceInfo.FOREGROUND_SERVICE_TYPE_CONNECTED_DEVICE)
        } else {
            startForeground(NOTIFICATION_ID, notification)
        }
    }

    private fun buildNotification(title: String, content: String): Notification {
        val launchIntent = Intent(this, MainActivity::class.java)
        val flags = PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        val pi = PendingIntent.getActivity(this, 0, launchIntent, flags)

        val builder = Notification.Builder(this, CHANNEL_ID)

        builder.setContentTitle(title)
            .setContentText(content)
            .setSmallIcon(android.R.drawable.ic_dialog_map)
            .setContentIntent(pi)
            .setOngoing(true)

        return builder.build()
    }

    private fun updateNotification() {
        val title = if (isConnected) "Pulsar NS400Z: Connected" else "Pulsar NS400Z: Disconnected"
        var content = "Nav: $activeManeuver ($activeStreet)"
        if (mediaListener?.isPlaying == true) {
            content += " | ♫ ${mediaListener?.currentTitle}"
        }

        val nm = getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager
        nm?.notify(NOTIFICATION_ID, buildNotification(title, content))
    }

    fun updateNavStatus(maneuver: String, street: String) {
        this.activeManeuver = maneuver
        this.activeStreet = street
        updateNotification()
    }

    override fun onConnectionStateChanged(connected: Boolean, deviceName: String, deviceAddress: String) {
        this.isConnected = connected
        updateNotification()

        if (connected) {
            mediaListener?.syncMetadata()
        }
    }

    override fun onPacketSent(charUuid: String, frame: ByteArray?, success: Boolean) {
        // Telemetry/TBT dispatched
    }

    override fun onHandlebarEvent(event: PulsarProtocol.HandlebarEvent) {
        Log.i(TAG, "Handlebar switch event received from cluster!")
        mediaListener?.handleHandlebarMedia(event)
        callHandler?.handleHandlebarCallAction(event)
    }

    fun refreshMediaSessions() {
        mediaListener?.initMediaSessions()
    }

    override fun onDestroy() {
        super.onDestroy()
        instance = null
        Log.i(TAG, "PulsarForegroundService onDestroy()")
        phoneMonitor?.stop()
        bleManager?.apply {
            removeListener(this@PulsarForegroundService)
            setAutoReconnect(false)
            stopScan()
            disconnect()
        }
        try {
            val nm = getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager
            nm?.cancel(NOTIFICATION_ID)
            nm?.cancelAll()
        } catch (ignored: Exception) {}
    }

    override fun onBind(intent: Intent?): IBinder? = null

    companion object {
        private const val TAG = "PulsarForegroundService"
        const val CHANNEL_ID: String = "pulsar_connect_live_channel"
        const val NOTIFICATION_ID: Int = 4001

        const val ACTION_START: String = "com.bajaj.rideconnect.re.START_SERVICE"
        const val ACTION_STOP: String = "com.bajaj.rideconnect.re.STOP_SERVICE"
        const val ACTION_CONNECT: String = "com.bajaj.rideconnect.re.CONNECT_BLE"

        @Volatile
        @JvmStatic
        var instance: PulsarForegroundService? = null
            private set

        @JvmStatic
        fun start(context: Context) {
            val intent = Intent(context, PulsarForegroundService::class.java).apply {
                action = ACTION_START
            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                context.startForegroundService(intent)
            } else {
                context.startService(intent)
            }
        }
    }
}
